### PR TITLE
feat: [PIE-7172]: Support for multi-service environment in deployment

### PR DESCRIPTION
**Pull request by sahilhindwani-harness**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/42417)

### CODE REVIEW

File: 120-ng-manager/src/main/java/io/harness/ng/core/deploymentstage/DeploymentStageConfigResource.java [hunk 1]
```diff
@@ -146,40 +157,66 @@ private List<YamlNode> getStageYamlNodesFromStageGroup(YamlNode stageGroupNode)
...
+            .map(ParameterField::getValue)
+            .collect(Collectors.toList());
       }
     }
-    return EMPTY;
+    return new ArrayList<>();
   }
 
-  private String getServiceRef(DeploymentStageConfig deploymentStageConfig) {
+  private List<String> getServiceRef(DeploymentStageConfig deploymentStageConfig) {
```
Comment (yogesh-chauhan) [inline 120-ng-manager/src/main/java/io/harness/ng/core/deploymentstage/DeploymentStageConfigResource.java:192]: can we call it getServiceRefs(..)

File: 120-ng-manager/src/main/java/io/harness/ng/core/deploymentstage/DeploymentStageConfigResource.java [hunk 2]
```diff
@@ -83,18 +88,24 @@ public class DeploymentStageConfigResource {
...
-      serviceRef = getServiceRef(referredStage.getDeploymentStageConfig());
+      serviceRefs = getServiceRef(referredStage.getDeploymentStageConfig());
     } else {
-      serviceRef = getServiceRef(actualStage.getDeploymentStageConfig());
+      serviceRefs = getServiceRef(actualStage.getDeploymentStageConfig());
     }
-    if (isNotBlank(serviceRef)) {
-      CDStageMetaDataDTOBuilder.serviceRef(serviceRef);
+
+    final List<String> environmentRefs = getEnvironmentRef(actualStage.getDeploymentStageConfig());
```
Comment (yogesh-chauhan) [inline 120-ng-manager/src/main/java/io/harness/ng/core/deploymentstage/DeploymentStageConfigResource.java:96]: can we call this method getEnvironmentRefs(..)

**Review by vaibhav-si — DISMISSED** [review]
> Reviewed 4 of 5 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42417)* status: 5 file(s) to review. 1 of 2 approvals obtained.

**Review by yogesh-chauhan — CHANGES_REQUESTED** [review]

**Review by tchaurasiya — DISMISSED** [review]
> Reviewed 4 of 5 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42417)* status: :shipit: complete! ✓

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 2 of 5 files at r1, 1 of 1 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42417)* status: 2 file(s) to review. 1 of 2 approvals obtained.

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42417)* status: 3 file(s) to review. 1 of 2 approvals obtained.

**Review by tchaurasiya — APPROVED** [review]
> Reviewed 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42417)* status: :shipit: complete! ✓

**Review by brijeshdhakar — APPROVED** [review]
> Reviewed 1 of 5 files at r1, 1 of 1 files at r2, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42417)* status: :shipit: complete! ✓

**Comment by sahilhindwani-harness** [comment]
trigger messagecheck

**Pull request merged**